### PR TITLE
AT-29

### PR DESCRIPTION
**Pull request by AamirahV**
Problem: A user can navigate to /cohorts/ and see information about the cohort with an id of 3. 

Solution: Added cohortSize and leadCoach columns to the cohorts table so that it would be displayed on the screen via the api. Created a GET route to navigate to api/cohorts/*id*. Created cypress tests to show that the correct information is passed through to the individual cohort pages.

Jira Issue: https://dfacademy.atlassian.net/jira/software/projects/AT/boards/4?selectedIssue=AT-29

### CODE REVIEW

**Comment by dearshrewdwit** [comment]
@AamirahV You need to update this branch

File: routes/cohorts.js [hunk 1]
```diff
@@ -11,12 +11,25 @@ router.get('/', async function (req, res) {
...
   }).then(cohorts =>{
-    res.json({ cohorts: cohorts })
+    res.json({ cohorts: cohorts})
   }).catch(errors => {
     res.json({ errors: ["Sorry invalid query parameters"] })
   })
 })
 
+router.get('/:id', async function (req, res) {
+  const students = await Student.findAndCountAll({
```
Comment (dearshrewdwit) [inline routes/cohorts.js:21]: Can you make one query instead? https://sequelize.org/master/manual/eager-loading.html#ordering-eager-loaded-associations

Something like:
```js
const cohort = await Cohort.findOne({
  where: { id: req.params.id },
  include: { all: true },
  order: [Student, 'firstName', 'ASC']
})
```
You should be able to access the data you need, and just send back the response as
(abridged for demo purposes)
```json
{ 
  "cohort": {
    "id":  4,
    "name": "mse-2106-a",
    "students": [
      {
         "id": 3,
         "firstName": "aaa"
      },
    ]
  }
}
```

**Review by dearshrewdwit — CHANGES_REQUESTED** [review]
Nearly there.

1 see comment about the sequelize query.
2. Why is there work on this PR that has already been merged?

File: routes/cohorts.js [hunk 2]
```diff
@@ -11,12 +11,26 @@ router.get("/", async function (req, res) {
...
-    res.json({ cohorts: cohorts });
-  }).catch(() => {
+    res.json({ cohorts: cohorts});
+  }).catch(()=> {
     res.json({ errors: ["Sorry invalid query parameters"] });
   });
 });
 
+router.get("/:id", async function (req, res) {
+  const cohort = await Cohort.findAndCountAll({
```
Comment (dearshrewdwit) [inline routes/cohorts.js:21]: the var name represents a set of cohorts - so let's name it `cohorts`

File: routes/cohorts.js [hunk 3]
```diff
@@ -11,12 +11,26 @@ router.get("/", async function (req, res) {
...
+    include: [{
+      model: Student},
```
Comment (dearshrewdwit) [inline routes/cohorts.js:23]: Why specify `model: Student`, when you also specify  `{ all: true }`

It's like having a bowl of apples, oranges, bananas, and saying can you get me all the fruit and the bananas!

File: routes/cohorts.js [hunk 4]
```diff
@@ -11,12 +11,26 @@ router.get("/", async function (req, res) {
...
 });
 
+router.get("/:id", async function (req, res) {
+  const cohort = await Cohort.findAndCountAll({
+    include: [{
+      model: Student}, 
+    { all: true 
+    }],
+    where: {
+      id: req.params.id},
```
Comment (dearshrewdwit) [inline routes/cohorts.js:27]: the linebreaks are out of sync here

**Review by dearshrewdwit — CHANGES_REQUESTED** [review]
Nice. Some minor changes.

1. What happens to the data already in production database where the cohorts don't have a lead instructor?

File: routes/cohorts.js (hunk 2 above)
Comment (dearshrewdwit) [reply routes/cohorts.js:21]: Is there a reason to use `findAndCountAll` over `findAll` ?

**Comment by AamirahV** [comment]
> Nice. Some minor changes.
> 
> 1. What happens to the data already in production database where the cohorts don't have a lead instructor?

returns null atm

File: routes/cohorts.js (hunk 2 above)
Comment (AamirahV) [reply routes/cohorts.js:21]: the count all returns a student count

**Comment by dearshrewdwit** [comment]
> > Nice. Some minor changes.
> > 
> > 1. What happens to the data already in production database where the cohorts don't have a lead instructor?
> 
> returns null atm

To avoid this, write a second migration to add a lead instructor to the cohorts that already exist so we avoid null fields -> which will cause problems if the frontend expects that data to exist.

Does that make sense?

File: routes/cohorts.js (hunk 3 above)
Comment (AamirahV) [reply routes/cohorts.js:23]: sorry i forgot to take it out as i was trying some different methods

**Comment by AamirahV** [comment]
> > > Nice. Some minor changes.
> > > 
> > > 1. What happens to the data already in production database where the cohorts don't have a lead instructor?
> > 
> > 
> > returns null atm
> 
> To avoid this, write a second migration to add a lead instructor to the cohorts that already exist so we avoid null fields -> which will cause problems if the frontend expects that data to exist.
> 
> Does that make sense?

sort of. so i would be assigning a default lead coach to the cohorts without a lead instructor?

**Comment by dearshrewdwit** [comment]
> > > > Nice. Some minor changes.
> > > > 
> > > > 1. What happens to the data already in production database where the cohorts don't have a lead instructor?
> > > 
> > > 
> > > returns null atm
> > 
> > 
> > To avoid this, write a second migration to add a lead instructor to the cohorts that already exist so we avoid null fields -> which will cause problems if the frontend expects that data to exist.
> > Does that make sense?
> 
> sort of. so i would be assigning a default lead coach to the cohorts without a lead instructor?

In a real example, you would have to find the correct coach name for the correct cohort. So let's simulate it.

There are 2 cohorts currently:

```sh
 id |    name    |           createdAt           |           updatedAt           |       startDate        
----+------------+-------------------------------+-------------------------------+------------------------
  2 | mse-2103-a | 2021-08-23 12:36:15.197566+00 | 2021-08-23 12:36:15.197566+00 | 2021-03-01 00:00:00+00
  3 | mse-2106-a | 2021-08-23 12:36:49.156668+00 | 2021-08-23 12:36:49.156668+00 | 2021-06-28 00:00:00+00
```
for cohort id 2, add a lead coach name: 'Edward Withers'
for cohort id 3, add a lead coach name: 'Edward's evil twin'

**Review by dearshrewdwit — COMMENTED** [review]
Fix style rules - then should be good to go

**Pull request closed without merging**